### PR TITLE
Improve in-app purchase for MacOS

### PR DESCRIPTION
**Pull request by AdrienFery**
This pull request adds multiple improvements of the In-App Purchase module for MacOS (see #11292).

1. It is now possible to get product descriptions (name, price, etc.) directly from the App Store.
2. It is now possible to finish transactions.
3. The documentation has been improved.
4. I have added a tutorial that shows how to use the In-App Purchase module. 

Let me know if you see modifications to do.

### CODE REVIEW

**Comment by AdrienFery** [comment]
Regarding the results of the automatic checks:

1. The linter fails but it doesn't seem to be related to In-App Purchase.
2. Regarding the amr64 build error, does the conversion of `std::vector<long long>` is not possible in arm64?

**Comment by codebytere** [comment]
@AdrienFery the linter dies on:
```
atom/browser/mac/in_app_purchase_product.h:25:  Use int16/int64/etc, rather than the C type long  [runtime/int] [4]
```

File: docs/tutorial/in-app-purchases.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+# In-App Purchase (macOS)
+
+## Preparing
+
+### Paid Applications Agreement
+If it's not already done, you’ll need to sign the Paid Applications Agreement and set up your banking and tax information in iTunes Connect.
```
Comment (zeke) [inline docs/tutorial/in-app-purchases.md]: "If it's not already done," might read more smoothly as "If you haven't already, "

File: docs/tutorial/in-app-purchases.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+### Create Your In-App Purchases
+Then, you'll need to configure your in-app purchases in iTunes Connect, and include details such as name, pricing, and description that highlights the features and functionality of your in-app purchase.
+
+[iTunes Connect Developer Help: Create an in-app purchase](https://help.apple.com/itunes-connect/developer/#/devae49fb316)
+
+
+### Change the CFBundleIdentifier
+
+To test In-App Purchase in developement with Electron you'll have to change the `CFBundleIdentifier` in `node_modules/electron/dist/Electron.app/Contents/Info.plist`. You have to replace `com.github.electron` by the budle identifier of the application you created with iTunes Connect.
```
Comment (zeke) [inline docs/tutorial/in-app-purchases.md]: budle -> bundle

File: docs/tutorial/in-app-purchases.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+```xml
+<key>CFBundleIdentifier</key>
+<string>com.example.app</string>
+```
+
+
+## Code example
+
+Here is an example that shows how tu use In-App Purchase in Electron. You'll have to replace the product ids by the identifiers of the products created with iTunes Connect (the identifier of `com.example.app.product1` is `product1`). Note that you have to listen to the `transactions-updated` event as soon as possible in your app.
```
Comment (zeke) [inline docs/tutorial/in-app-purchases.md]: tu -> to

Comment (zeke) [reply docs/tutorial/in-app-purchases.md]: Purchase -> Purchases

File: spec/api-in-app-purchase-spec.js [hunk 4]
```diff
@@ -32,4 +40,11 @@ describe('inAppPurchase module', function () {
       done()
     })
   })
+
+  it('getProducts() return an empty list when getting invalid product', (done) => {
```
Comment (zeke) [inline spec/api-in-app-purchase-spec.js]: return -> returns

**Comment by zeke** [comment]
I can't speak to the implementation, but the docs look ready to go. 👍

**Comment by yuri-karkh** [comment]
@AdrienFery I got an edge case with this pull request: if bundle ID contains "-" like we already have and what is allowed by Apple, then we can't point product just by ID like u mentioned: (The identifier of `com.example.app.product1` is `product1`), because product ID is not allowed to contain "-".

**Comment by AdrienFery** [comment]
@yuri-karkh Did you change the CFBundleIdentifier in the `Info.plist` as explained in the [tutorial](https://github.com/AdrienFery/electron/blob/39a9b51ecc143295c13a9788e0ffc62b40ad8571/docs/tutorial/in-app-purchases.md#change-the-cfbundleidentifier) ?

**Comment by yuri-karkh** [comment]
Yes @AdrienFery  and I got login popup so all good, but I mean little different thing. I mean that is I use real bundle which ID contains "-", then product ID will be not valid, e.g. if bundle ID = com.test.app-test then product ID com.test.app-test.product1 will not be valid.

**Comment by AdrienFery** [comment]
@yuri-karkh 

> The identifier of com.example.app.product1 is product1

What I mean is that the user shouldn't include the bundleID in the productID. In the example, the product ID is `product1` and not `com.test.app-test.product1`.

**Comment by yuri-karkh** [comment]
Sorry maybe I didn't get this from start but, don't you mean that product id that we crate in itunes should be like bundleID.productID and simply productID @AdrienFery ? And then in the code u just join bundleID and productID?

**Comment by yuri-karkh** [comment]
hi @AdrienFery sorry to bother u again, but have a question again in u have a minute:
I use electron@2.0.0-beta.7, I have changed Info.plist to my bundle id (double checked) and when I calling inAppPurchase.productPurchase nothing happens now. Also tested canMakePayments, getReceiptURL methods they look ok. No errors in console.

code
console

**Comment by immanoj16** [comment]
Hey @AdrienFery, According to the tutorial this path `node_modules/electron/dist/Electron.app/Contents/Info.plist` is not present in the current beta version. Can you help me. Thank you

**Comment by immanoj16** [comment]
Hey @AdrienFery Can you please tell how it works in mac app store means how it will work in electron app itself. Please make a clear documentation or tutorial. Thank you

**Comment by zeke** [comment]
@immanoj16 there is a tutorial here: https://github.com/electron/electron/blob/master/docs/tutorial/in-app-purchases.md -- if you find issues with it, please open a new issue or pull request to address them.

**Comment by immanoj16** [comment]
@zeke Thank you

**Comment by JeganMSD** [comment]
@AdrienFery While calling purchase product App getting crashed on Mac version <10.10.... Why???

**Comment by JeganMSD** [comment]
@AdrienFery or anyone. Why getProducts method not working even in Electron 2.0.6.But working beta-3.0.1... When this method made legally to work..
Please correct me If I am wrong..

**Comment by omar10594** [comment]
We need these features on v2.x

**Comment by JeganMSD** [comment]
@omar10594  me too looking forward for this feature. As this feature is still in Beta version 😞 . I had my belief on @AdrienFery for the faster implementation on stable version.

**Comment by holgersindbaek** [comment]
@AdrienFery Are you planning on implementing `restoreCompletedTransactions ` (https://stackoverflow.com/questions/7738120/what-does-restore-purchases-in-in-app-purchases-mean#7738220), so users can "properly" restore their purchases?

Right now I'm calling `purchaseProduct`  when people click "Restore purchase", which will restore users purchase if they've purchased before, but it also shows people the normal purchase popup, which makes a lot of users think that they have to buy the game again.

**Pull request merged**